### ISSUE TITLE
[Bug Report]InvalidSignatureException Invalid signature has given

### ISSUE THREAD

**Issue by nabejohn** [line/line-bot-sdk-php#269]
## Bug Report

**Describe the bug**
receive message with emoji from messaging API.
`InvalidSignatureException Invalid signature has given` occurs.
message without emoji is no InvalidSignatureException occurs.

**To Reproduce**
1. Send message from LINE iOS App  'テスト😭'
2. Receive message on messaging API line-bot-sdk-php 
3. `LINEBot->parseEventRequest()`
4. `InvalidSignatureException Invalid signature has given `

**Expected behavior**
No InvalidSignatureException occurs.

**Screenshots**
![line-error](https://user-images.githubusercontent.com/8167667/72787606-9829c000-3c73-11ea-8f74-f5171d962ad4.png)

**Environment**
 - OS: Amazon Linux 2018.03
 - PHP version  5.5.38
 - line-bot-sdk-php version 4.1.0

**Additional context**
I reported line-community too.
https://www.line-community.me/question/5e26b241851f7402cd950561/invalidsignatureexception-invalid-signature-has-given

**Comment by tokuhirom** [comment]
`InvalidSignatureException Invalid signature has given` means, your LINEBot doesn't have the correct channel secret. You need to check the constructor argument of `\LINE\LINEBot`.

The example code is here.

```
$httpClient = new \LINE\LINEBot\HTTPClient\CurlHTTPClient('<channel access token>');
$bot = new \LINE\LINEBot($httpClient, ['channelSecret' => '<channel secret>']);
```

So, you can get your channel secret from [the LINE Developer Center](https://developers.line.biz/console/).

**Comment by nabejohn** [comment]
@tokuhirom thanks reply!
message without emoji `テスト` is no InvalidSignatureException occurs.
and postback message no InvalidSignatureException occurs too.

maybe i have correct channel secret..

**Comment by tokuhirom** [comment]
hmm...  Maybe, there is an issue around `$events = $bot->parseEventRequest($body, $signature)`. 
How do you get `$body` in your code?

**Comment by nabejohn** [comment]
variable $body print to log file like this.
I masked some value like this `[REPLY_TOKEN]`.

- `テスト` no InvalidSignatureException occurs $body.
```
{
	"events": [{
		"type": "message",
		"replyToken": "[REPLY_TOKEN]",
		"source": {
			"userId": "[USER_ID]",
			"type": "user"
		},
		"timestamp": 1579602735658,
		"mode": "active",
		"message": {
			"type": "text",
			"id": "[ID]",
			"text": "テスト"
		}
	}],
	"destination": "[DESTINATION]"
}
```
- `テスト😭` InvalidSignatureException occurs $body.
```
{
	"events": [{
		"type": "message",
		"replyToken": "[REPLY_TOKEN]",
		"source": {
			"userId": "[USER_ID]",
			"type": "user"
		},
		"timestamp": 1579602923647,
		"mode": "active",
		"message": {
			"type": "text",
			"id": "[ID]",
			"text": "テスト😭"
		}
	}],
	"destination": "[DESTINATION]"
}
```

type of $body is string.
```
echo gettype($body);
=> string
```

**Comment by moririnson** [comment]
@nabejohn Sorry for the inconvenience.
We use `hash_equals` for signature validate.
https://github.com/line/line-bot-sdk-php/blob/master/src/LINEBot/SignatureValidator.php#L45
PHP seems to implement `hash_equals` since 5.6.0. How do you implement `hash_equals` function?
> Environment
> PHP version 5.5.38
https://www.php.net/manual/ja/function.hash-equals.php
> (PHP 5 >= 5.6.0, PHP 7)

**Comment by nabejohn** [comment]
@moririnson thanks reply!
```
# php -r "echo function_exists('hash_equals') ? \"yes\n\" : \"no\n\";"
no
```

> How do you implement hash_equals function?
I use facebook/php-graph-sdk too.
maybe `hash_equals` implementation is below.
https://github.com/facebook/php-graph-sdk/blob/5.x/src/Facebook/polyfills.php

**Comment by moririnson** [comment]
@nabejohn 
Thanks. I couldn't find any problems that function...

Could you print the body by `file_get_contents('php://input')` ? 
If you get `テスト😭` instead of` テスト\uD83D\uDE2D`, you may have any problems with the server Unicode escape as you said.

**Comment by moririnson** [comment]
@nabejohn 
Thanks. I couldn't find any problems that function...

Could you print the body by file_get_contents('php://input') ? 
And compared to body that passed to `Event RequestParser#parseEventRequest`.
If you get a diff, you probably have a problem with your Request parse.

**Comment by nabejohn** [comment]
@tokuhirom @moririnson 
thanks for many advice !

I found a solution.
The cause was the conversion of emoji.
It is not SDK bug . so sorry.

`😭` surrogate pairs `\ud83d\ude2d` is InvalidSignatureException occurs.
`😭` surrogate pairs `\uD83D\uDE2D` is no InvalidSignatureException occurs.

I use `json_decode()` and `json_encode()` .
```
echo "😭 ==> " . json_encode("😭");
😭 ==> "\ud83d\ude2d"
```

so I implemented converter (`\ud83d\ude2d` => `\uD83D\uDE2D`). and problem is gone.
